spacemesh/v2alpha1/tx.proto

```diff
@@ -59,4 +59,106 @@ message TransactionV1 {
 
 message Nonce {
   uint64 counter = 1;
+}
+
+// TransactionState is the "journey" of a tx from mempool to block inclusion to
+// mesh to STF processing. To know whether or not the tx actually succeeded,
+// and its side effects, check the Receipt in the GlobalStateService.
+message TransactionState {
+  bytes id = 1;
+  enum TransactionState {
+    TRANSACTION_STATE_UNSPECIFIED = 0; // default state
+    TRANSACTION_STATE_REJECTED = 1; // rejected from mempool due to, e.g., invalid syntax
+    TRANSACTION_STATE_INSUFFICIENT_FUNDS = 2; // rejected from mempool by funds check
+    TRANSACTION_STATE_CONFLICTING = 3; // rejected from mempool due to conflicting counter
+    TRANSACTION_STATE_MEMPOOL = 4; // in mempool but not on the mesh yet
+    TRANSACTION_STATE_MESH = 5; // submitted to the mesh
+    TRANSACTION_STATE_PROCESSED = 6; // processed by STF; check Receipt for success or failure
+  }
+  TransactionState state = 2;
+}
+
+message TransactionResult {
+  enum Status {
+    SUCCESS = 0;
+    FAILURE = 1;
+    INVALID = 2;
+  }
+
+  Transaction tx = 1;
+  Status status = 2;
+  string message = 3;
+  uint64 gas_consumed = 4; // in units of gas
+  uint64 fee = 5; // in smidge
+  bytes block = 6;
+  uint32 layer = 7;
+  repeated string touched_addresses = 8;
+}
+
+message TransactionStreamRequest {
+  uint32 start_layer = 1;
+  uint32 end_layer = 2;
+  repeated bytes id = 3;
+  string address = 4;
+  bool watch = 5;
+}
+
+service TransactionStreamService {
+  rpc Stream(TransactionStreamRequest) returns (stream Transaction);
+}
+
+message TransactionRequest {
+  uint32 start_layer = 1;
+  uint32 end_layer = 2;
+  repeated bytes id = 3;
+  string address = 4;
```

Comment:
What is `address` here? If it means principal we should rename to `principal`. Or does it mean any tx that touches an account? Let's also please add an explanatory comment.

Reply:
I think we need to be more specific than "involved in." What does that mean precisely? Right now we could limit it to principal and recipient of a send, plus maybe spawn destination. In the future if we add more functionality at L1 this might get more complicated, or even be infeasible (e.g., there's basically no way for an Ethereum node to know which accounts were involved in a historical ERC-20 transaction, and it would be the same for some account state update that happened inside a tx at L2 in our case).

My inclination is to remove this for now. Let's just allow a request using `principal` since it should be easier to index on that. Maybe we can allow `recipient` of a coin send as well.

Do we know how we're going to implement this?